REGRESSION (UI-side compositing) overflow:scroll inside SVG foreignObject fails to render scrollbars

nmoucht opened the pull request:
#### 4f1ded96b3967cf2de4cc260b1e49940e93822e4
<pre>
REGRESSION (UI-side compositing) overflow:scroll inside SVG foreignObject fails to render scrollbars
<a href="https://bugs.webkit.org/show_bug.cgi?id=261227">https://bugs.webkit.org/show_bug.cgi?id=261227</a>
<a href="https://rdar.apple.com/115075206">rdar://115075206</a>

Reviewed by Simon Fraser.

Keep NSScrollerImps in the web process for scrollers inside svg foreignObject as
they don&apos;t follow the normal code path for creating scrollbars in the UI process.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::usesOverlayScrollbars const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):
* Source/WebCore/platform/mac/NSScrollerImpDetails.h:
* Source/WebCore/platform/mac/NSScrollerImpDetails.mm:
(WebCore::ScrollerStyle::usesOverlayScrollbars):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::registerScrollbarsForScrollableAreaInSVGForeignObject):
(WebCore::RenderLayer::updateAncestorDependentState):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::isInsideSVGForeignObject const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::shouldRegisterScrollbars const):

Canonical link: <a href="https://commits.webkit.org/277015@main">https://commits.webkit.org/277015@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af85545053aa0c6b734d576bb0c24603b885e994

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25446 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48983 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42351 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48617 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29802 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22903 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37825 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46888 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22510 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39937 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19050 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19883 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41032 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4354 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42605 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41387 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50796 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17786 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45025 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22606 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40089 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43942 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22973 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6476 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22303 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -246,6 +246,8 @@ class RenderLayer : public CanMakeWeakPtr<RenderLayer> {
     // Convert a point in absolute coords into layer coords, taking transforms into account
     LayoutPoint absoluteToContents(const LayoutPoint&) const;
 
+    bool isInsideSVGForeignObject() const { return m_insideSVGForeignObject; }
```
I'd prefer to avoid this reference to SVG foreignObject here. It should be more generic, like `needsToRenderOwnScrollbars` or something.

nmoucht replied:
This function was already on `RenderLayer` (just made it public) so its probably ok to keep.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.h:
```diff
@@ -256,6 +256,8 @@ class RenderLayerScrollableArea final : public ScrollableArea {
     void updateScrollPositionForScrollAnchoringController() final;
     void invalidateScrollAnchoringElement() final;
+    void createScrollbarsController() final;
+    void registerScrollbarsWhenInSVGForeignObject();
```
Same comment on naming here.

nmoucht replied:
Updated to `registerScrollbars()`.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -2048,5 +2050,17 @@ void RenderLayerScrollableArea::invalidateScrollAnchoringElement()
 
+void RenderLayerScrollableArea::registerScrollbars()
+{
+    createScrollbarsController();
```
We should make this safe to call more than once (i.e. add `ensureScrollbarsController()`).

nmoucht replied:
I added an explanation of the solution to the changelog but we need to change the ScrollbarsController type if we are going to keep the scrollbars in the web process, so this should remain a call to `createScrollbarsController`.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -2048,5 +2050,17 @@ void RenderLayerScrollableArea::invalidateScrollAnchoringElement()
+{
+    createScrollbarsController();
+    if (auto* horizontalScrollbar = this->horizontalScrollbar()) {
+        horizontalScrollbar->theme().registerScrollbar(*horizontalScrollbar);
```
ScrollBar's constructor calls `theme().registerScrollbar(*this);` so why do we do this again here?

nmoucht replied:
Same thing as above we need to re-register the scrollbars when we change the ScrollbarsController type.

smfr replied:
I think what's confusing is that the roles of `ScrollbarTheme` and `ScrollbarsController` are not really clear. Scrollbars register themselves with the theme at creation; registering them again with (possibly a different) theme here seems error prone. Maybe `ScrollbarsController` should be responsible for registering a scrollbar with the theme?

Also, if the `ScollbarsController` can change, I'd prefer to see an explicit "null out the old one, make a new one" code pattern.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
```diff
@@ -96,7 +97,10 @@
 
 bool RemoteScrollbarsController::shouldRegisterScrollbars() const
```
We should invent a name for "having scrollbars that need to render in the web process" and use it here and instead of exposing `isInsideSVGForeignObject` in more places.

nmoucht replied:
Changed this to scrollable area having `shouldRenderScollbarsInProcess`.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -2052,5 +2054,23 @@ void RenderLayerScrollableArea::invalidateScrollAnchoringElement()
+
+bool RenderLayerScrollableArea::shouldRenderScollbarsInProcess()
+{
+    return m_layer.isInsideSVGForeignObject();
```
Isn't it really about not being able to use asyncScrolling, i.e it will have no ScrollingNodeID? So is this the same as when `RenderLayerScrollableArea::usesAsyncScrolling()` returns false? Or perhaps `RenderLayerScrollableArea::canUseCompositedScrolling()` ?

nmoucht replied:
Right, using `RenderLayerScrollableArea::usesAsyncScrolling` simplifies the patch quite a bit since we don't have to set it from `RenderLayer::updateAncestorDependentState`, we can just query `RenderLayerScrollableArea::usesAsyncScrolling` at scrollbar controller creation time.

smfr reviewed (approved):
Can we make a test?

nmoucht committed:
REGRESSION (UI-side compositing) overflow:scroll inside SVG foreignObject fails to render scrollbars
https://bugs.webkit.org/show_bug.cgi?id=261227
rdar://115075206

Reviewed by Simon Fraser.

Keep NSScrollerImps in the web process for scrollers inside svg foreignObject as
they don't follow the normal code path for creating scrollbars in the UI process.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::usesOverlayScrollbars const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):
* Source/WebCore/platform/mac/NSScrollerImpDetails.h:
* Source/WebCore/platform/mac/NSScrollerImpDetails.mm:
(WebCore::ScrollerStyle::usesOverlayScrollbars):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::registerScrollbarsForScrollableAreaInSVGForeignObject):
(WebCore::RenderLayer::updateAncestorDependentState):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::isInsideSVGForeignObject const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::shouldRegisterScrollbars const):

Canonical link: https://commits.webkit.org/277015@main
